extra_data/reader.py

```diff
@@ -592,18 +592,41 @@ def union(self, *others):
         """
         files = set(self.files)
         train_ids = set(self.train_ids)
+        is_single_run = self.is_single_run
+        run_num = None
+        proposal_num = None
+
+        # Account for old runs (file format 0.5),
+        # which doesn't have these information.
+        if (
+            "runNumber" in self.run_metadata().keys() and
+            "proposalNumber" in self.run_metadata().keys()
+        ):
+            if self.is_single_run:
+                run_num = self.run_metadata()["runNumber"]
+                proposal_num = self.run_metadata()["proposalNumber"]
 
         for other in others:
             files.update(other.files)
             train_ids.update(other.train_ids)
+            # Set other.is_single_run to False
+            # in case of Unioning multiple run DataCollections.
+
+            if not (
+                other.is_single_run and
+                other.run_metadata()["runNumber"] == run_num and
+                other.run_metadata()["proposalNumber"] == proposal_num
```

Comment:
You've handled if `self` comes from an older file without this metadata, but you're still assuming that `other` has these details.